mir_eval/sonify.py

```diff
@@ -81,36 +82,56 @@ def time_frequency(gram, frequencies, times, fs, function=np.sin, length=None):
     Returns
     -------
     output : np.ndarray
-        synthetized version of the piano roll
+        synthesized version of the piano roll
 
     """
     # Default value for length
+    if times.ndim == 1:
+        # Convert to intervals
+        times = util.boundaries_to_intervals(times)
+
     if length is None:
-        length = int(times[-1]*fs)
+        length = int(times[-1, 1] * fs)
+
+    times, _ = util.adjust_intervals(times, t_max=length)
+
+    # Truncate times so that the shape matches gram
+    times = times[:gram.shape[1]]
 
     def _fast_synthesize(frequency):
-        """A faster (approximate) way to synthesize a signal
-            synthesize a few periods then repeat that signal
+        """A faster way to synthesize a signal.
+            Generate one cycle, and simulate arbitrary repetitions
+            using array indexing tricks.
         """
-        # Generate ten periods at this frequency
-        ten_periods = int(10*fs*(1./frequency))
-        short_signal = function(2*np.pi*np.arange(ten_periods)*frequency/fs)
-        # Repeat the signal until it's of the desired length
+        # Generate one period at this frequency
+        short_signal = function(np.linspace(0, 2 * np.pi,
+                                            num=fs/float(frequency),
+                                            endpoint=False))
```

Comment:
Ok, don't mean to get super pedantic here, but when you're synthesizing a single period and repeating it, the effective frequency of your signal is `fs/int(fs/frequency)`.  For low frequencies, this isn't too different from `frequency`; for high frequencies it can be very different.  For example, for MIDI note 90 (~1480 hz) with `fs = 22050`, `fs/int(fs/frequency)` is 1575 Hz.  That's why I synthesized ten periods; this was a quick hack to make the effective frequency `10*fs/int(10*fs/frequency)` which for MIDI note 90 results in about 1490 Hz.  It seemed like a good trade-off as an alternative to adding some logic to keep track of the sample error when you're repeating `short_signal`.  I'm open to ideas if you have a better one (one would be to use a very-fast-to-compute pure Python triangle wave implementation or something, and don't do any repeat tricks).  For reference, here are plots of the semitone error of all MIDI notes from 0 to 110 (assuming notes above 110 are a wash) for repeating a single period vs. repeating 10 periods:
```Python
fs = 22050.
def plot(errs):
    plt.figure(figsize=(15, 8))
    plt.xticks(np.arange(0, 128, 4))
    plt.plot(errs)
    plt.grid()
    plt.xlabel('MIDI Note')
    plt.ylabel('Semitone error')
plot([librosa.hz_to_midi(fs/int(fs/librosa.midi_to_hz(f))) - f
      for f in np.arange(0, 110)])
plt.title('1 period')
plot([librosa.hz_to_midi(10*fs/int(10*fs/librosa.midi_to_hz(f))) - f
      for f in np.arange(0, 110)])
plt.title('10 periods')
```
![download 4](https://cloud.githubusercontent.com/assets/417568/12892231/2932ce6c-ce59-11e5-84d3-9b9db7d05905.png)
![download 7](https://cloud.githubusercontent.com/assets/417568/12892239/384e31c0-ce59-11e5-836c-ec341f9e1f1c.png)